rllib/env/wrappers/atari_wrappers.py

```diff
@@ -208,28 +208,60 @@ def reset(self, **kwargs):
 
 @PublicAPI
 class MaxAndSkipEnv(gym.Wrapper):
-    def __init__(self, env, skip=4):
-        """Return only every `skip`-th frame"""
-        gym.Wrapper.__init__(self, env)
-        # most recent raw observations (for max pooling across time steps)
-        self._obs_buffer = np.zeros((2,) + env.observation_space.shape, dtype=np.uint8)
+    """Performs frame skipping on env with max'ing observations over last 2 frames.
+
+    Note that when using this wrapper with Atari envs, your `frameskip` gym.make()
+    setting should probably be set to 1, as you probably don't want to have two
+    frame skipping wrappers active.
+
+    Within the skipped frames, this wrapper will return the maximum individual-dim
+    observation values over the last two frames.
+    For example, if observation_space = Box(0, 255, (1, ), uint8) and we skip over the
+    4 observations: [9] [4] [100] and [59], we would return [100] (max over the last
+    two skipped frames).
+    """
+    def __init__(self, env, skip: int = 4, max_over: int = 2):
+        """Initializes a MaxAndSkipEnv instance.
+
+        Args:
+            skip: The number of frames to skip with each `step()` call. Note that
+                after `reset()` no frames are skipped.
+            max_over: The number of the last n frames (within the skipped ones) that
+                we max the returned observations over. Each dim in the observation
+                space is maxed over individually, not changing the shape (or dtype)
+                of the observation space.
+        """
+        super().__init__(env)
+
+        assert skip >= max_over, (
+            f"`skip` ({skip}) value must be larger or equal to `max_over` "
+            f"({max_over}) value!"
+        )
         self._skip = skip
+        self._max_over = max_over
+
+        # Most recent raw observations (for max pooling across last two time steps).
+        self._obs_buffer = np.zeros(
+            (self._max_over,) + env.observation_space.shape,
+            dtype=env.observation_space.dtype,
+        )
 
     def step(self, action):
         """Repeat action, sum reward, and max over last observations."""
         total_reward = 0.0
         terminated = truncated = info = None
         for i in range(self._skip):
             obs, reward, terminated, truncated, info = self.env.step(action)
-            if i == self._skip - 2:
-                self._obs_buffer[0] = obs
-            if i == self._skip - 1:
-                self._obs_buffer[1] = obs
+            # If self._skip == 4, buffer_idx moves from [3 .. 0]
+            buffer_idx = self._skip - i - 1
```

Comment:
Does it matter if we insert from the back of the buffer is bigger than skip? Are we supposed to be shifting elements from right to left as we insert, that way we retain old obs data? If so let's use a dequeue. Otherwise let's always make the buffer of size skip.

Reply:
Good question:
It doesn't matter, from which side we insert into the buffer as we are doing max'ing at the end.
Also, we always completely fill the buffer from beginning to end within a single `step()` b/c we skip at least as many frames as we max over.
We should not make the buffer of size skip as `skip` is always >= `max_over`. The max'd over steps are a subset of the skipped steps, so if skip=4 and max_over=2 then we throw away the first 2 skipped obs completely (w/o ever looking at them) and only use the 3rd and 4th obs to max over and return that.